### PR TITLE
pin dependency to a branch

### LINKED ISSUE #2211: Pin a specific dependency version
### Component

Forge

### Describe the feature you would like

# Issue
Currently, when installing a specific version/branch of a dependency, this version is lost when running `forge update`.
Example:
```
forge install rari-capital/solmate@v7
forge update
```
=> checks out to master
I would expect the checked out branch to remain the same.

# Example Scenario
I currently am depending on solmate@v7's LibString.sol and have multiple of my own dependencies in my projects, which I frequently update. Every time I want to update, I currently run:
```
forge update
forge remove solmate
forge install rari-capital/solmate@v7
```

I realize that I also could just update the packages individually. Still, I would expect forge not to switch branches when updating.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dsam82**
## Motivation

Resolves #2211

## Solution

Utilises `git submodules set-branch` feature to pin a submodule to a branch. 

Adds a `-b` option to `forge install` to specify the branch, otherwise looks for tag or commit hash.

Would love to know if there is any other better solution possible.

## Example

```bash
lforge install -b rari-capital/fuse-v1@feat/
Installing fuse-v1 in "/Users/sambhavdusad/Projects/testt/lib/fuse-v1" (url: Some("https://github.com/rari-capital/fuse-v1"), tag: Some("feat/"))
There are multiple matching branches:
[0] feat/ (original branch)
[1] feat/contract-packages
[2] feat/add-vulnerability-analysis
[3] feat/add-liquidation-tests
[4] feat/add-deployment-test-flow
Please select a tag (0-4, default: 1): 1
[1] feat/contract-packages selected
    Installed fuse-v1
```

Installs fuse-v1 and pin it to branch chosen by user.

### Disclaimer

This is my first PR in Rust, so there might have been absolute horrible things that I've done. Looking forward for your reviews :P

### CODE REVIEW

**Comment by onbjerg** [comment]
The @ syntax we already have supports branches, so I think the primary issue in #2211 is not that you can't install a specific branch, it's that `forge update` ignores it. Any thoughts on how we might solve that instead? Imo (given @ supports branches already) a `-b` flag does not make sense

**Comment by dsam82** [comment]
`-b` flag sets the submodule to point to a particular branch using [git submodule set-branch](https://git-scm.com/docs/git-submodule#Documentation/git-submodule.txt-set-branch-b--branchltbranchgt--ltpathgt) option.

Updates the `.gitmodules`

```bash
[submodule "lib/fuse-v1"]
        path = lib/fuse-v1
        url = https://github.com/rari-capital/fuse-v1
        branch = feat/contract-packages
```

Now, `forge update` won't update to remote **HEAD** but to that specific **branch**

I added `-b` so that we can know that user wants to pin a branch, if he wants specific tag or commit also without forge updating it, then I think `--no-git` is the best way.

Do you have any other solution in mind?

**Comment by onbjerg** [comment]
I would rather we autodetect if it is a branch then instead of adding a new flag, should be doable. How would you update to a new branch if the branch is pinned also?

**Comment by dsam82** [comment]
> I would rather we autodetect if it is a branch then instead of adding a new flag, should be doable.

So, we should first detect for branch and then for tags?

> How would you update to a new branch if the branch is pinned also?

Currently, no other way through forge, you can directly use submodules command.

```
git submodules add-branch -b <branch> <path/alias>
```

**Comment by onbjerg** [comment]
@dsam82 I think we need to address moving to a different branch/tag as well in this PR otherwise the behavior might be a bit confusing

>So, we should first detect for branch and then for tags? I thought there may be some branch or tags with same name and detecting automatically might create confusions, so created a flag.

I think that's preferable, right now if a branch and a tag are named the same thing the behavior is also somewhat undefined. I also *think* (but am not entirely sure) that if a branch and a tag is named the same thing, we can expect the tag to point to the latest commit on the branch with the same name eventually, so it should be OK

**Comment by dsam82** [comment]
@onbjerg removed `-b` flag, and we can change branch after pinning through `forge install`.

```bash
forge install foundry-rs/forge-std@docs-fix
```

Changes branch to `docs-fix` even if it's pinned to another branch and updates `branch` tag in `.gitmodules`, then running `forge update` updates to that branch.

File: cli/src/cmd/forge/install.rs [hunk 1]
```diff
@@ -148,27 +163,31 @@ fn install_as_submodule(
...
-        format!("forge install: {target_dir}\n\n{tag}")
+        Some(tag)
     };
 
     // commit the added submodule
     if !no_commit {
+        let message = match &tag {
+            Some(i) => format!("forge install: {target_dir}\n\n{i}"),
+            None => format!("forge install: {target_dir}"),
+        };
```
Comment (mattsse) [inline cli/src/cmd/forge/install.rs:186]: use `if let Some()` syntax for Options

File: cli/src/cmd/forge/install.rs [hunk 2]
```diff
@@ -341,19 +371,91 @@ fn match_tag(tag: &String, libs: &Path, target_dir: &str) -> eyre::Result<String
...
+    // fetch remote branches and check for tag
+    let output = Command::new("git")
+        .args(&["branch", "-r"])
+        .current_dir(&libs.join(&target_dir))
+        .get_stdout_lossy()?;
+    let mut candidates: Vec<String> = output
+        .split('\n')
+        .map(|x| x.trim().trim_start_matches("origin/"))
+        .filter(|x| x.starts_with(&tag.to_string()))
+        .map(|x| x.to_string())
```
Comment (mattsse) [inline cli/src/cmd/forge/install.rs:401]:
```suggestion
        .map(str::to_string)
```

**Review by mattsse — CHANGES_REQUESTED** [review]
couple smol nits

also needs nightly formatting `cargo +nightly fmt`

defer to @onbjerg for final approval

**Review by onbjerg — APPROVED** [review]
LGTM

**Pull request merged**